overlord/devicestate/systems.go

```diff
@@ -176,3 +186,183 @@ func seededSystemFromModeenv() (*seededSystem, error) {
 	}
 	return seededSys, nil
 }
+
+func createSystemForModelFromValidatedSnaps(getInfo func(name string) (*snap.Info, bool, error), db asserts.RODatabase, label string, model *asserts.Model) (newFiles []string, dir string, err error) {
+	if isUC20 := model.Grade() != asserts.ModelGradeUnset; !isUC20 {
+		return nil, "", fmt.Errorf("cannot create a system for non UC20 model")
+	}
+
+	logger.Noticef("creating recovery system with label %q for %q", label, model.Model())
+
+	wOpts := &seedwriter.Options{
+		// RW mount of ubuntu-seed
+		SeedDir: boot.InitramfsUbuntuSeedDir,
+		Label:   label,
+	}
+	w, err := seedwriter.New(model, wOpts)
+	if err != nil {
+		return nil, "", err
+	}
+
+	optsSnaps := make([]*seedwriter.OptionsSnap, 0, len(model.RequiredWithEssentialSnaps()))
+
+	// collect all snaps that are present
+	modelSnaps := make(map[string]*snap.Info)
+	for _, sn := range model.EssentialSnaps() {
+		info, present, err := getInfo(sn.SnapName())
+		if err != nil {
+			return nil, "", fmt.Errorf("cannot obtain snap information: %v", err)
+		}
+		// grab those
+		logger.Noticef("essential snap: %v", sn.SnapName())
+		if !present {
+			return nil, "", fmt.Errorf("internal error: essential snap %q not present", sn.SnapName())
+		}
+		// present locally
+		optsSnaps = append(optsSnaps, &seedwriter.OptionsSnap{
+			Path: info.MountFile(),
+		})
+		modelSnaps[info.MountFile()] = info
+	}
+	for _, sn := range model.SnapsWithoutEssential() {
+		info, present, err := getInfo(sn.SnapName())
+		if err != nil {
+			return nil, "", fmt.Errorf("cannot obtain non-essential snap information: %v", err)
+		}
+		logger.Noticef("non-essential but %q snap %v", sn.Presence, sn.SnapName())
+		if sn.Presence == "optional" && !present {
+			// the snap is optional
+			continue
+		}
+		if !present {
+			return nil, "", fmt.Errorf("internal error: non-essential but %q snap %q not present", sn.Presence, sn.SnapName())
+		}
+		// present locally
+		optsSnaps = append(optsSnaps, &seedwriter.OptionsSnap{
+			Path: info.MountFile(),
+		})
+		modelSnaps[info.MountFile()] = info
+	}
+	if err := w.SetOptionsSnaps(optsSnaps); err != nil {
+		return nil, "", err
+	}
+
+	newFetcher := func(save func(asserts.Assertion) error) asserts.Fetcher {
+		fromDB := func(ref *asserts.Ref) (asserts.Assertion, error) {
+			return ref.Resolve(db.Find)
+		}
+		return asserts.NewFetcher(db, fromDB, save)
+	}
+	f, err := w.Start(db, newFetcher)
+	if err != nil {
+		return nil, "", err
+	}
+
+	localSnaps, err := w.LocalSnaps()
+	if err != nil {
+		return nil, "", err
+	}
+
+	for _, sn := range localSnaps {
+		// TODO: is the side info here different from what we have in snap.Info?
+		_, aRefs, err := seedwriter.DeriveSideInfo(sn.Path, f, db)
+		if err != nil && !asserts.IsNotFound(err) {
```

Comment:
if info will came from the state (I suppose) and info.SnapID != ""  then a assertions not found is a real error that we shouldn't ignore, I suppose something later might check for that though? anyway we should double check and have a comment here about that

Reply:
tweaked this to catch a case when we have a snap ID but no assertion